server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,292 @@
+        return this.isAtLeast(IBP_3_0_IV0);
+    }
+
+    public boolean isTruncationOnFetchSupported() {
+        return this.isAtLeast(IBP_2_7_IV1);
+    }
+
+    public boolean isOffsetForLeaderEpochSupported() {
+        return this.isAtLeast(IBP_0_11_0_IV2);
+    }
+
+    public boolean isTopicIdsSupported() {
+        return this.isAtLeast(IBP_2_8_IV0);
+    }
+
+    public boolean isFeatureVersioningSupported() {
+        return this.isAtLeast(IBP_2_7_IV1);
+    }
+
+    public boolean isSaslInterBrokerHandshakeRequestEnabled() {
+        return this.isAtLeast(IBP_0_10_0_IV1);
+    }
+
+    public RecordVersion recordVersion() {
+        if (this.compareTo(IBP_0_9_0) <= 0) { // IBPs up to IBP_0_9_0 use Record Version V0
+            return RecordVersion.V0;
+        } else if (this.compareTo(IBP_0_10_2_IV0) <= 0) { // IBPs up to IBP_0_10_2_IV0 use V1
+            return RecordVersion.V1;
+        } else return RecordVersion.V2; // all greater IBPs use V2
+    }
+
+    private static final Map<String, MetadataVersion> IBP_VERSIONS;
+    static {
+        {
+            IBP_VERSIONS = new HashMap<>();
+            Pattern versionPattern = Pattern.compile("^IBP_([\\d_]+)(?:IV(\\d))?");
+            Map<String, MetadataVersion> maxInterVersion = new HashMap<>();
+            for (MetadataVersion version : MetadataVersion.values()) {
+                Matcher matcher = versionPattern.matcher(version.name());
+                if (matcher.find()) {
+                    String withoutIV = matcher.group(1);
+                    // remove any trailing underscores
+                    if (withoutIV.endsWith("_")) {
+                        withoutIV = withoutIV.substring(0, withoutIV.length() - 1);
+                    }
+                    String shortVersion = withoutIV.replace("_", ".");
+
+                    String normalizedVersion;
+                    if (matcher.group(2) != null) {
+                        normalizedVersion = String.format("%s-IV%s", shortVersion, matcher.group(2));
+                    } else {
+                        normalizedVersion = shortVersion;
+                    }
+                    maxInterVersion.compute(shortVersion, (__, currentVersion) -> {
+                        if (currentVersion == null) {
+                            return version;
+                        } else if (version.compareTo(currentVersion) > 0) {
+                            return version;
+                        } else {
+                            return currentVersion;
+                        }
+                    });
+                    IBP_VERSIONS.put(normalizedVersion, version);
+                } else {
+                    throw new IllegalArgumentException("Metadata version: " + version.name() + " does not fit "
+                            + "any of the accepted patterns.");
+                }
+            }
+            IBP_VERSIONS.putAll(maxInterVersion);
+        }
+    }
+
+    public String shortVersion() {
```

Comment:
Can we move the logic of `shortVersion` and `version` into the constructor (or in the static initializer)? These values won't change for the lifetime of MetadataVersion instance, so we can make them final fields that we calculate once.